sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs

```diff
@@ -0,0 +1,238 @@
+            {
+                OnSendingRequest(message);
+            }
+
+            try
+            {
+                if (async)
+                {
+                    await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+                }
+                else
+                {
+                    ProcessNext(message, pipeline, currentIndex);
+                }
+            }
+            catch (Exception ex)
+            {
+                allTryExceptions ??= new List<Exception>();
+                allTryExceptions.Add(ex);
+
+                thisTryException = ex;
+            }
+
+            if (async)
+            {
+                await OnRequestSentAsync(message).ConfigureAwait(false);
+            }
+            else
+            {
+                OnRequestSent(message);
+            }
+
+            bool shouldRetry = async ?
+                await ShouldRetryAsync(message, thisTryException).ConfigureAwait(false) :
+                ShouldRetry(message, thisTryException);
+
+            if (shouldRetry)
+            {
+                TimeSpan delay = GetNextDelay(message, message.RetryCount);
+                if (delay > TimeSpan.Zero)
+                {
+                    if (async)
+                    {
+                        await WaitAsync(delay, message.CancellationToken).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        Wait(delay, message.CancellationToken);
+                    }
+                }
+
+                // Dispose the content stream to free up a connection if the request has any
+                message.Response?.ContentStream?.Dispose();
+
+                message.RetryCount++;
+                OnTryComplete(message);
+
+                continue;
+            }
+
+            if (thisTryException != null)
+            {
+                // Rethrow if there's only one exception.
+                if (allTryExceptions!.Count == 1)
+                {
+                    ExceptionDispatchInfo.Capture(thisTryException).Throw();
+                }
+
+                throw new AggregateException($"Retry failed after {message.RetryCount + 1} tries.", allTryExceptions);
+            }
+
+            // ShouldRetry returned false this iteration and the last request
+            // we sent didn't cause an exception to be thrown.
+            // So, we're done.  Exit the while loop.
+            break;
+        }
+    }
+
+    protected virtual void OnSendingRequest(PipelineMessage message) { }
+
+    protected virtual ValueTask OnSendingRequestAsync(PipelineMessage message) => default;
+
+    protected virtual void OnRequestSent(PipelineMessage message) { }
+
+    protected virtual ValueTask OnRequestSentAsync(PipelineMessage message) => default;
+
+    protected virtual void OnTryComplete(PipelineMessage message) { }
+
+    internal bool ShouldRetry(PipelineMessage message, Exception? exception)
```

Comment:
nit: combine ShouldRetry and ShouldRetryAsync into something like ShouldRetrySyncOrAsync.

Reply:
Addressed